### PR TITLE
[experiment] os_thread_io_priority_class and os_thread_io_priority_level (ionice for threads)

### PR DESCRIPTION
**Pull request by filimonov**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- New Feature

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
per-thread ionice (same way as os_thread_priority)

Detailed description / Documentation draft:
See
https://man7.org/linux/man-pages/man2/ioprio_set.2.html 
https://www.kernel.org/doc/Documentation/block/ioprio.txt

### CODE REVIEW

File: src/Core/SettingsCollection.h [hunk 1]
```diff
@@ -295,6 +295,16 @@ enum class DistributedProductMode
 };
 using SettingDistributedProductMode = SettingEnum<DistributedProductMode>;
 
+/// The setting for executing distributed subqueries inside IN or JOIN sections.
```
Comment (alexey-milovidov) [inline src/Core/SettingsCollection.h:298]: Misleading comment.

**Review by alexey-milovidov — COMMENTED** [review]
ionice is set automatically with nice - maybe we don't need a separate setting.

**Comment by alexey-milovidov** [comment]
What IO schedulers support priority class?

**Comment by filimonov** [comment]
> What IO schedulers support priority class? 

CFQ only afaik.

> ionice is set automatically with nice - maybe we don't need a separate setting.

That's why '[experiment]'.  

> ionice is set automatically with nice

is true inside 'best_effort' io_priority_class.  You can't use 'realtime' & 'idle' by adjusting nice.

**Comment by alexey-milovidov** [comment]
Ok, I cannot wait to see some test results...

**Comment by alexey-milovidov** [comment]
Did the experiment go well?

**Comment by filimonov** [comment]
Sorry, it looks like i will not have time for finishing that this week, and next 2 weeks I'm on vacation, so if someone want to peek it in the meantime - it's ok for me (i can back to that in 3-4 weeks). 

Some general thoughts:
1) realtime - clickhouse will have higher priority for accessing the disk, it's risk-prone is it's a single disk used also for os. It may be used if some disk is heavily used by lot of other threads / processes and we need to take some data bypassing the main queue. I have serious doubts if we should make that avaliable for users (there are quite high risk of misusing that). 
2) best_effort - is used by default, and priority is adjusted based on nice with formula  `io_nice = (cpu_nice + 20) / 5`. But it not always good to scale them together. For example i need to read a data without disturbing access to disk of other queries, but want to compress the result very well (http, gzip). If i will set nice to 19 the thread will get scheduled very rate comparing to other threads, i.e. even if it gets the data from disk it will not get enough cpu time to compress that for sending (it can lead to connection timeouts etc). So setting low io_priority in that case sounds much better (anyway it will stay inactive if it waits fot data from socket). So main usecase is long-running export-like `selects`  which can CPU when they they get some data to process, but should stay in queue if some other queries require cpu. 
To test that you need to run 2 concurrent loads:
```
curl + gzip - long running, io intensive select, max_threads=1 
clickhouse-benchmark with some queries creating moderate cpu & disk load (directio to avoid caches).
```
And compare it's the behavior when for first case nice is adjusted to values like 19, with the case when ionice for the first query is set to 7.
3) 'idle' will get access to disk ONLY if nobody from higher io class needs that. That can be nice for some maintenance operations - like ttl drop, move parts between disks, some backfilling (insert ... select ...) etc. The difference with os_thread_io_priority_level = 7 in that os_thread_io_priority_level = 7 should still get some (tiny) part of io bandwidth even if disk is loaded with some higher priority tasks, with 'idle' it will just wait.

Test is something like 
```
INSERT INTO ... SELECT ...  -- io & cpu intensive
clickhouse-benchmark with some queries creating moderate cpu & disk load (directio to avoid caches). 
```
My expectations is that the first query with nice = 19 will still disturb the io access of the clickhouse-benchmark queries, and with idle it will wait till clickhouse-benchmark queries will finish.

**Comment by alexey-milovidov** [comment]
@filimonov Ok, we need to show some results of experiments to prove that this makes sense.

**Comment by alexey-milovidov** [comment]
Please make a new PR if we will continue.

**Pull request closed without merging**